#3129 - Domain Renewal Form (Expiring Soon and Expired) - [AD]

therealslimhsiehdy opened the pull request:
## Ticket

Resolves #3129

## Changes

** [Previous PR ](https://github.com/cisagov/manage.get.gov/pull/3248)

* [Domain Renewal Form Figma](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar)
* [Domain Renewal Detail Page Expired](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=14310-15859&p=f)
* For expiring soon and expired domains, we now have a renewal form for domain managers to fill out and/or confirm through 
* Documentation on how to add in an expiring soon domain to play with on sandbox

## Context for reviewers

For domains that are expired and expiring soon, we now have a Renewal form that domain managers can fill out to renew their domains by 1 year. 

## Setup

**EXPIRING DOMAINS in Org Model Mode** under the `Hotel California`:

```
cooscountyor.gov
detroit.gov
elkrivermn.gov
gloucesterva.gov
headoftheharborny.gov
healthdata.gov
hvlnc.gov
jeffersonknoxohio.gov
montgomerycountyhealthky.gov
pembinewi.gov
pikecoga.gov
richlandcountywi.gov
stantonca.gov
texasjcmh.gov
texasstateparks.gov
vanburencountytn.gov
ipageaz.gov
```

**EXPIRING DOMAINS in Non Org Mode:**
NOTE:
You will need to go into domain, search for domain -> manage domain -> add self as domain manager
If your email is not working, add yourself as an allowed email list on my sandbox

```
arcoidaho.gov
hendersoncountytn.gov
hinsdalema.gov
linndalevillage-oh.gov
mapletonmn.gov
mountvernonin.gov
neenahwi.gov
newmadridcountymo.gov
nls.gov
pointpleasantbeachnj.gov
swcleanair.gov
waitehilloh.gov
hardwick-ma.gov
```

---------------------------------------------------------------------------------------------------------------------------------------

**EXPIRED DOMAINS in Org Model Mode** under the `Hotel California portfolio`:
```
erieco.gov
hendersoncountyil.gov
humboldt-wi.gov
lafayettecountymo.gov
laquintaca.gov
legmt.gov
ngla.gov
shreveportlawater.gov
somersettx.gov
stalbanswv.gov
swaincountync.gov
texicopolicenm.gov
build.gov
kingstonsprings-tn.gov
maizeks.gov
pitu.gov
randallso.gov
townofnilesny.gov
woodhavenmi.gov
```

**EXPIRED DOMAINS** in **In Non Org Mode:**
NOTE:
You will need to go into domain, search for domain -> manage domain -> add self as domain manager
If your email is not working, add yourself as an allowed email list on my sandbox

```
bearcreektownshipmi.gov
brantleycounty-ga.gov
chesapeakewv.gov
fairmontnc.gov
fultoncountyny.gov
lptami.gov
montgomerycountync.gov
opensource.gov
peoriaparks-il.gov
riverwoods.gov
santaclarita.gov
scit-nsn.gov
townofdunbarwi.gov
warnernh.gov
waynecountyil.gov
flpd6.gov
```

1. In theory, I should've added enough domains that are expiring soon to the RH sandbox -- if you do not see any, please let either Aditi and/or Rebecca know and we can add some in for you (sorry for the inconvenience!). 
2. Go to `/admin` and turn the `domain renewal` waffle flag on (no org model yet)
3. If you are not a domain manager of any of these domains, add yourself as a domain manager (via /admin -> Domains -> Manage Domains -> Add yourself)
4. You should now see the renewal form in the sidebar. You can also access the form on the Domain Overview page via the link `Renew to maintain access` under the status. 
5. When you arrive at the Domain Renewal Form, you can test out editing your contact, security, and domain managers. Then, submit it to renew the domain. 
6. Now go to `/admin` and turn on the `organization_feature` flag and run through the same scenarios as above. (The ones I have added should be in the `Agency Of The Litterbox Association` portfolio, and you will probably need to add yourself as a domain manager to play around with it)

**Note:** The only difference between the org model and non org model view is that the org model view has breadcrumbs on top, but the non org model view does. The form should otherwise be the same. 

## Code Review Verification Steps

### As the original developer, I have

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Update documentation in READMEs and/or onboarding guide

#### Ensured code standards are met (Original Developer)
<!-- Mark "- N/A" and check at the end of each check that is not applicable to your PR -->
- [ ] If any updated dependencies on Pipfile, also update dependencies in requirements.txt.
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] Tag @dotgov-designers in this PR's Reviewers for design review. If code is not user-facing, delete design reviewer checklist
- [ ] Verify new pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

### As a code reviewer, I have

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Verified code meets all checks above. Address any checks that are not satisfied
- [ ] Reviewed this code and left comments. Indicate if comments must be addressed before code is merged
- [ ] Checked that all code is adequately covered by tests
- [ ] Verify migrations are valid and do not conflict with existing migrations

#### Validated user-facing changes as a developer
**Note:** Multiple code reviewers can share the checklists above, a second reviewer should not make a duplicate checklist. All checks should be checked before approving, even those labeled N/A. 

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] (Rarely needed) Tested as both an analyst and applicant user

### As a designer reviewer, I have

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior. Comment any found issues or broken flows.
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari
- [ ] (Rarely needed) Tested as both an analyst and applicant user

### References
- [Code review best practices](../docs/dev-practices/code_review.md)

## Screenshots

<!-- If this PR makes visible interface changes, an image of the finished interface can help reviewers
and casual observers understand the context of the changes.
A before image is optional and can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow.
You may want to use [GIPHY Capture](https://giphy.com/apps/giphycapture) for this! 📸

_Please frame images to show useful context but also highlight the affected regions._
--->

abroddrick commented on docs/developer/registry-access.md:
```diff
@@ -103,3 +103,50 @@ response = registry._client.transport.receive()
+If you want to be in the org model mode, turn the `organization_feature` waffle flag on, and add that domain via Django Admin to a portfolio to be able to view it. 
+
+
+
```
**nitpick** can we remove all the extra endlines?

therealslimhsiehdy replied:
Omg not a nitpick at all, idk what happened here sorry -- great catch!

abroddrick commented on src/registrar/tests/test_views_domain.py:
```diff
@@ -453,12 +454,21 @@ def setUp(self):
 
         self.user.save()
 
+    def todays_expiration_date(self):
```
**suggestion** Maybe rename this function, it doesn't have return todays expiration date but an expiration date 1 year in the future

abroddrick commented on src/registrar/models/domain.py:
```diff
@@ -328,7 +328,6 @@ def renew_domain(self, length: int = 1, unit: epp.Unit = epp.Unit.YEAR):
             # if no expiration date from registry, set it to today
             logger.warning("current expiration date not set; setting to today")
             exp_date = date.today()
-
```
**minor changes requested**: I see the renew code is logging errors below here but it is not printing the full traceback or the domain that threw the error. Please change both the RegistryError and normal exception to include traceback and domain name.

therealslimhsiehdy replied:
Would you mind elaborating or give an example of what you mean by traceback (I'm looking into our code for a pattern, but want to make sure I'm referencing the right area 😅)?

abroddrick replied:
Oh no problem you just want to add exec_info=true on the logger line! [docs](https://docs.python.org/3/library/logging.html#logging.Logger.debug)
example from our code:
 `           logger.warning("Failed to send confirmation email", exc_info=True)`

asaki222 replied:
Updated!

abroddrick commented on src/registrar/views/domain.py:
```diff
@@ -311,6 +311,85 @@ def _get_domain(self, request):
+                domain.renew_domain()
+                messages.success(request, "This domain has been renewed for one year.")
+            except Exception as e:
+                messages.error(request, "This domain has not been renewed for one year, error was %s" % e)
```
You don't want to tell the user what the error was unless it helps them solve the problem on their own or will help them communicate the problem to us (such as an error code). I think here you would want to tell them to email help@get.gov if this problem persists. That way they have a clear action they can take.

abroddrick commented on src/registrar/views/domain.py:
```diff
@@ -311,6 +311,85 @@ def _get_domain(self, request):
+
+        # Check if the checkbox is checked
+        is_policy_acknowledged = request.POST.get("is_policy_acknowledged", None)
+        if is_policy_acknowledged != "on":
```
**question:** This is different than how we usually control validation, for every other page we use the form structure and have made sure to keep that pretty standard in the registrar. Why not change this to be a form with this as it's field so that it is checked on is_valid?

kristinacyin commented:
As discussed in #dotgov-design-reviews, I will do the first high-level pass review and another designer will take a look after. Deadline for PR review is Friday, 1/10.

kristinacyin commented:
Umm...tried to review but the domain is "Ready" even though the expiration date is today? 
![image](https://github.com/user-attachments/assets/8ef03865-9e46-41b4-82f1-53158c488cd9)

kristinacyin reviewed (changes requested):
Amazing work! Noticed a few little discrepancies between the design and the implementation: 

- There should be 16px space below the H3 headers ("Your contact information", "Security email", and "Domain managers." 
- It also seems like the spacing between "None provided" text in the "Security email" section seems to be too big, compared to other sections? The spacing between the sections should be 32px. 
- The "Your contact information" section header should be in sentence case ("contact" and "information" should not be capitalized)

| Figma Design  | Implementation |
| ------------- | ------------- |
| ![tempsnip](https://github.com/user-attachments/assets/5173a336-518c-4655-ad26-f3afa5fc1960) | ![image](https://github.com/user-attachments/assets/2e42d943-b1ce-4f3c-a126-6187a521180a)  |

- It seems like the text for the "Expired" status alert hasn't been updated yet (has "Expiring soon" text)

| Figma Design  | Implementation |
| ------------- | ------------- |
| ![image](https://github.com/user-attachments/assets/05ac052b-948c-4edc-ba68-950a8e435625) | ![image](https://github.com/user-attachments/assets/62b2ec97-47d8-4889-9185-c760c31d73f4) |

**Potential bugs**
- With org model turned on, even though I didn't have any "Expiring soon" domains, I still saw the "Multiple domains are expiring" banner notification:
![image](https://github.com/user-attachments/assets/9a06297b-4663-4497-ae92-42c01f11a8d8)
- For some reason when I tried to update the security email, the Renewal Form will always show the Security email as "None provided"

therealslimhsiehdy commented:
@kristinacyin Thank you so much for the thorough review from the screenshots to the descriptions!

Below are responses to your comments:

> There should be 16px space below the H3 headers ("Your contact information", "Security email", and "Domain managers."

A small pushback - I matched the pattern to what we have on Domain Overview (so no spacing between the header and the text below) - although the design may look better with the spacing, code wise it would be a bit of a large override to match the Figma design so I did not make the spacing change.

> It also seems like the spacing between "None provided" text in the "Security email" section seems to be too big, compared to other sections? The spacing between the sections should be 32px.

Fixed! GREAT catch. Turns out it was a rogue p tag.

> The "Your contact information" section header should be in sentence case ("contact" and "information" should not be capitalized)

Fixed! 

> It seems like the text for the "Expired" status alert hasn't been updated yet (has "Expiring soon" text)

Fixed!

> The "Expiring soon" status is missing in the filter list options when the user doesn't have any "Expiring soon" domains. Given that the other status options are still shown even if the user doesn't have domains with all the different statuses, "Expiring soon" should still be shown regardless

Fixed!

> For some reason when I tried to update the security email, the Renewal Form will always show the Security email as "None provided"

Fixed -- huge thank you for catching this, was a backend code issue.

> There might also be a weird error where even though I checked the acknowledgement checkbox of the renewal form, it will still give me the error message as if I haven't. But it works when I check & submit the second time

Should be fixed - I think my sandbox is borked - please use Adit's `ad` sandbox for now. I will update the description and such to match that and add expiring and expired domains to her sandbox.

> With org model turned on, even though I didn't have any "Expiring soon" domains, I still saw the "Multiple domains are expiring" banner notification:

This is the only one I wasn't sure of -- could you happen to give me steps to replicate this? Not sure if this was an org model bug or my sandbox being a butthead bug.

dave-kennedy-ecs commented on src/registrar/views/domain.py:
```diff
@@ -311,6 +311,89 @@ def _get_domain(self, request):
+                    return True
+        return False
+
+    def in_editable_state(self, pk):
+        """Override in_editable_state from DomainPermission
+        Allow detail page to be viewable"""
+
+        requested_domain = None
+        if Domain.objects.filter(id=pk).exists():
+            requested_domain = Domain.objects.get(id=pk)
+
+        # return true if the domain exists, this will allow the detail page to load
+        if requested_domain:
+            return True
+        return False
```
If I read the code correctly, this method is inherited. Is that correct? Looks like an exact copy of the inherited method. If I am reading correctly, this method should not be redefined here.

dave-kennedy-ecs commented on src/registrar/views/domain.py:
```diff
@@ -311,6 +311,89 @@ def _get_domain(self, request):
+
+    def _get_domain(self, request):
+        """
+        override get_domain for this view so that domain overview
+        always resets the cache for the domain object
+        """
+        self.session = request.session
```
I am confused by the purpose of this override. Are you intending to reset the domain in cache each time the renewal form is requested?

dave-kennedy-ecs commented on src/registrar/views/domain.py:
```diff
@@ -311,6 +311,89 @@ def _get_domain(self, request):
         self._update_session_with_domain()
 
 
+class DomainRenewalView(DomainBaseView):
```
Curious why inherit DomainBaseView instead of DomainView? It looks like 4 of the 5 methods in DomainRenewalView are copied from DomainView. Couldn't you inherit DomainView, and then just define the template and post method?

asaki222 replied:
@dave-kennedy-ecs That is a good point! the `__get__domain` functionality that we used is also available in `DomainView`. I just pushed up the changes!

dave-kennedy-ecs reviewed (dismissed):
I left some comments/questions on the view. Will re-evaluate once those are addressed.

kristinacyin commented:
> @kristinacyin Thank you so much for the thorough review from the screenshots to the descriptions!
> 
> Below are responses to your comments:
> 
> > There should be 16px space below the H3 headers ("Your contact information", "Security email", and "Domain managers."
> 
> A small pushback - I matched the pattern to what we have on Domain Overview (so no spacing between the header and the text below) - although the design may look better with the spacing, code wise it would be a bit of a large override to match the Figma design so I did not make the spacing change.
> 
> > It also seems like the spacing between "None provided" text in the "Security email" section seems to be too big, compared to other sections? The spacing between the sections should be 32px.
> 
> Fixed! GREAT catch. Turns out it was a rogue p tag.
> 
> > The "Your contact information" section header should be in sentence case ("contact" and "information" should not be capitalized)
> 
> Fixed!
> 
> > It seems like the text for the "Expired" status alert hasn't been updated yet (has "Expiring soon" text)
> 
> Fixed!
> 
> > The "Expiring soon" status is missing in the filter list options when the user doesn't have any "Expiring soon" domains. Given that the other status options are still shown even if the user doesn't have domains with all the different statuses, "Expiring soon" should still be shown regardless
> 
> Fixed!
> 
> > For some reason when I tried to update the security email, the Renewal Form will always show the Security email as "None provided"
> 
> Fixed -- huge thank you for catching this, was a backend code issue.
> 
> > There might also be a weird error where even though I checked the acknowledgement checkbox of the renewal form, it will still give me the error message as if I haven't. But it works when I check & submit the second time
> 
> Should be fixed - I think my sandbox is borked - please use Adit's `ad` sandbox for now. I will update the description and such to match that and add expiring and expired domains to her sandbox.
> 
> > With org model turned on, even though I didn't have any "Expiring soon" domains, I still saw the "Multiple domains are expiring" banner notification:
> 
> This is the only one I wasn't sure of -- could you happen to give me steps to replicate this? Not sure if this was an org model bug or my sandbox being a butthead bug.

kristinacyin reviewed (approved):
There's a couple of things that need to be relegated to post-POD work, but great job! Can't believe we've accomplished so much in such a short amount of time.

dave-kennedy-ecs commented on src/registrar/views/domain.py:
```diff
@@ -311,6 +311,59 @@ def _get_domain(self, request):
+
+    template_name = "domain_renewal.html"
+
+    def can_access_domain_via_portfolio(self, pk):
+        """Most views should not allow permission to portfolio users.
+        If particular views allow permissions, they will need to override
+        this function."""
+        portfolio = self.request.session.get("portfolio")
+        if self.request.user.has_any_domains_portfolio_permission(portfolio):
+            if Domain.objects.filter(id=pk).exists():
+                domain = Domain.objects.get(id=pk)
+                if domain.domain_info.portfolio == portfolio:
+                    return True
+        return False
```
One more. Looks like this one should also be deleted.

dave-kennedy-ecs reviewed (approved):
Approve, pending removal of the additional noted method in the view.

Thank you for all of the test cases in the PR notes. This was especially helpful for testing, and Rebecca was really helpful jumping on a call to walk through the PR
